src/charm.py

```diff
@@ -43,7 +46,16 @@ def __init__(self, *args: Any) -> None:
         except CharmConfigInvalidError as exc:
             self.model.unit.status = ops.BlockedStatus(exc.msg)
             return
-        self._synapse = Synapse(charm_state=self._charm_state)
+        self._database = DatabaseObserver(self)
+        self.framework.observe(
+            self._database.database.on.database_created, self._on_database_created
+        )
+        self.framework.observe(
+            self._database.database.on.endpoints_changed, self._on_config_changed
+        )
+        self._synapse = Synapse(
+            charm_state=self._charm_state, database_data=self._database.get_relation_data()
```

Comment:
Should the database relation data be passed via state?

Reply:
Sure, changed.